jax/interpreters/masking.py

```diff
@@ -104,30 +106,98 @@ class ShapeExpr(tuple):  # type ShapeExpr = [Poly]
   def __str__(self):
     return 'ShapeExpr({})'.format(', '.join(map(str, self)))
   def __getitem__(self, idx):
-    if type(idx) is int:
-      return super(ShapeExpr, self).__getitem__(idx)
-    else:
-      return ShapeExpr(super(ShapeExpr, self).__getitem__(idx))
+    poly = super(ShapeExpr, self).__getitem__(idx)
+    return poly if type(idx) is int else ShapeExpr(poly)
+
+
+def canonicalize_poly(p):
+  if onp.isscalar(p) and onp.issubdtype(onp.int, onp.array(p).dtype):
+    return constant_poly(p)
+
+  assert isinstance(p, Poly)
+  return p
+
+def canonicalize_polymorphic_shape(shape):
+  return ShapeExpr(map(canonicalize_poly, shape))
+
+def poly_without_zeros(d):
```

Comment:
Can you explain why you need this function? Is it for correctness or just efficiency? A comment could be a good tool here!

Reply:
This function is now removed.

Reply:
Actually, I was asking about `_poly_without_zeros`, though I realize now you may have been answering about `canonicalize_polymorphic_shape`.

Why do you need `_poly_without_zeros`? Just pruning / efficiency?

Reply:
Ah, I see. I was just fixing a bug related to this. I inlined `_poly_without_zeros` into the `Poly` constructor now and added a line to explain the purpose: It's to keep `Poly` in canonical form in order to simplify operations like comparison etc.

Reply:
Excellent, that makes sense! Thanks for explaining.